src/server/pps/server/api_server.go

```diff
@@ -502,6 +629,11 @@ func newPipelineInfo(persistPipelineInfo *persist.PipelineInfo) *ppsclient.Pipel
 func (a *apiServer) runPipeline(pipelineInfo *ppsclient.PipelineInfo) error {
 	ctx, cancel := context.WithCancel(context.Background())
 	a.cancelFuncsLock.Lock()
+	if _, ok := a.cancelFuncs[*pipelineInfo.Pipeline]; ok {
+		// The pipeline is already being run
+		a.cancelFuncsLock.Unlock()
+		return nil
+	}
```

Comment:
Yeah I haven't rigorous analyzed whether under some corner cases a pipeline can be run twice.  This is more of a safety measure.